First draft for text-to-image, image-to-image + generate script

Wauplin opened the pull request:
(related to https://github.com/huggingface/hub-docs/pull/1379)

cc @osanseviero for viz'

End goal is to generate this page based on info from:
- [input specs](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/tasks/text-to-image/spec/input.json)
- [output specs](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/tasks/text-to-image/spec/output.json)
- the task page https://huggingface.co/api/tasks
- (maybe `huggingface_hub` doc example?)
- (maybe `huggingface.js` doc example?)

I wrote the content in this PR manually to validate the format.  I have added:
- a small description from the tasks page
- a link to https://huggingface.co/tasks/text-to-image for more info
- a list of recommended models from the tasks page
- the API specification
    - inputs: payload (from specs) and headers (will always be the same)
    - output: payload (from specs). In the `text-to-image` example it's annoying because the output is not a json, no not describable using openschema. We should find a proxy to say "it's just bytes" in the specs so that docs are generated correctly.
- examples:
    - CURL => how to generate (?)
    - Python =>  from `hugginface_hub` example (?). Add link to docs.
    - JavaScript => from `huggingface.js` (?). Add link to docs.

Open questions:
- [x] we should add an example URL like `https://api-inference.huggingface.co/models/black-forest-labs/FLUX.1-dev`. Depends on what we do for the curl example but we need it in any case.
- [x] should we add a curl example? is it possible to generate it? are the current examples maintained?
- [x] we should harmonize python/javascript snippets with the ones from https://huggingface.co/black-forest-labs/FLUX.1-dev?inference_api=true? If yes, how?

Anything else?

osanseviero commented:
>  should we add a curl example? is it possible to generate it? are the current examples maintained?

Yes, adding a curl example is important and were recently updated :+1:

Wauplin commented:
I've updated the draft based on @osanseviero feedback:
- take snippets from https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/snippets => will be the same as for https://huggingface.co/black-forest-labs/FLUX.1-dev?inference_api=true
- added Javascript / cURL as well
- hard-coded small description (not taken from an API)
- added link for https://huggingface.co/models?inference=warm&pipeline_tag=text-to-image&sort=trending

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+
+- [black-forest-labs/FLUX.1-dev](https://huggingface.co/black-forest-labs/FLUX.1-dev): one of the most powerful image generation models that can generate realistic outputs.
+- [latent-consistency/lcm-lora-sdxl](https://huggingface.co/latent-consistency/lcm-lora-sdxl): a powerful yet fast image generation model.
+- [Kwai-Kolors/Kolors](https://huggingface.co/Kwai-Kolors/Kolors): text-to-image model for photorealistic generation.
```
This model is frozen. I think it's ok for now but let's consider filtering for only warm/cold models in the future

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+- [Kwai-Kolors/Kolors](https://huggingface.co/Kwai-Kolors/Kolors): text-to-image model for photorealistic generation.
+- [stabilityai/stable-diffusion-3-medium-diffusers](https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers): a powerful text-to-image model.
+
+This is only a subset of the supported models. Find the model that suits you best [here](https://huggingface.co/models?inference=warm&pipeline_tag=text-to-image&sort=trending).
```
Opened an internal issue so we can do an OR of warm and cold

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+| **inputs** | _string, required_ | The input text data (sometimes called "prompt"). |
+| **parameters.guidance_scale** | _number, optional_ | For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality. |
+| **parameters.negative_prompt[]** | _string, optional_ | FOne or several prompt to guide what NOT to include in image generation. |
+| **parameters.num_inference_steps** | _integer, optional_ | For diffusion models. The number of denoising steps. More denoising steps usually lead to a higher quality image at the expense of slower inference. |
```
I'm not a fan of the bunch of `parameters. ...`. Let's think if we can make something that keeps a clear difference while not being so repetitive

Wauplin replied:
for the record, I tried using nested lists

```
- **inputs** (_string, required_): The input text data (sometimes called "prompt").
- **parameters** (_object, optional_):
  - **guidance_scale** (_number, optional_): For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality.
  - **negative_prompt[]** (_string or string[], optional_): One or several prompts to guide what NOT to include in image generation.
  - **num_inference_steps** (_integer, optional_): For diffusion models. The number of denoising steps. More denoising steps usually lead to a higher quality image at the expense of slower inference.
  - **target_size** (_object, optional_):
    - **width** (_integer, optional_): The size in pixels of the output image.
    - **height** (_integer, optional_): The size in pixels of the output image.
  - **scheduler** (_string, optional_): For diffusion models. Override the scheduler with a compatible one.
```

resulting in 
![image](https://github.com/user-attachments/assets/17b2cc8c-589d-4950-a3ab-2c950113bef3)

Wauplin replied:
also tried nested table like this, which I'm really not a fan

```
| Payload |   |    |
| :--- | :--- | :--- | :--- | :--- |
| **inputs** |  |  | _string, required_ | The input text data (sometimes called "prompt"). |
| **parameters**  |  | |  |  |
|  | **guidance_scale**  || _number, optional_ | For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality. |
|  | **negative_prompt[]** | | _string, optional_ | One or several prompt to guide what NOT to include in image generation. |
|  | **num_inference_steps** | | _integer, optional_ | For diffusion models. The number of denoising steps. More denoising steps usually lead to a higher quality image at the expense of slower inference. |
|  | **target_size**  ||  |  |
|  |  | **width** | _integer, optional_ | The size in pixel of the output image. |
|  |  | **height** | _integer, optional_ | The size in pixel of the output image. |
|  | **scheduler**  | | _string, optional_ | For diffusion models. Override the scheduler with a compatible one. |
```

![image](https://github.com/user-attachments/assets/08d1e66c-1f92-4d0d-8088-b3fa4f345f18)

Wauplin replied:
Let's go with the last solution. It's ugly markdown-wise but the table looks ok. Can be changed in the future.

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+| :--- | :--- | :--- |
+| **inputs** | _string, required_ | The input text data (sometimes called "prompt"). |
+| **parameters.guidance_scale** | _number, optional_ | For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality. |
+| **parameters.negative_prompt[]** | _string, optional_ | FOne or several prompt to guide what NOT to include in image generation. |
```
The typing seems a bit off for me. Is it always an array of strings (even if it's just one?)

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+| **inputs** | _string, required_ | The input text data (sometimes called "prompt"). |
+| **parameters.guidance_scale** | _number, optional_ | For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality. |
+| **parameters.negative_prompt[]** | _string, optional_ | FOne or several prompt to guide what NOT to include in image generation. |
+| **parameters.num_inference_steps** | _integer, optional_ | For diffusion models. The number of denoising steps. More denoising steps usually lead to a higher quality image at the expense of slower inference. |
```
To check if all of these already work well out of the box

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+
+| Payload |   |    |
+| :--- | :--- | :--- |
+| **inputs** | _string, required_ | The input text data (sometimes called "prompt"). |
```
Do we want to specify limits? E.g `num_inference_steps` can't be larger than N, etc

Wauplin replied:
We could be it can get tricky. I think for a v1 it's fine not to have them.

We already discussed this when writing the specs. In theory it's possible to define lower and upper bound in a json schema. But in practice, the limits can depend on the model or even on the framework powering the model (example [here](https://github.com/huggingface/huggingface_hub/pull/1968#issuecomment-1903542670) with `top_p` that can be 1.0 in transformers but not in TGI). That being said, most integer/number values have fixed limits.

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+| **parameters.target_size.height** | _integer, optional_ | The size in pixel of the output image. |
+| **parameters.scheduler** | _string, optional_ | For diffusion models. Override the scheduler with a compatible one. |
+
+| Headers |   |    |
```
Sounds good, I would also document this in `docs/api-inference/task_parameters.md` maybe as it's a general parameter for all models

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
````diff
@@ -0,0 +1,103 @@
+
+#### Python
+
+```py
````
TODO in a follow-up: switch to the python library?

Wauplin replied:
Yes! To be done in https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/snippets/python.ts

osanseviero commented on docs/api-inference/tasks/text-to-image.md:
```diff
@@ -0,0 +1,103 @@
+	return response.content
+image_bytes = query({
+	"inputs": "Astronaut riding a horse",
+})
+# You can access the image with PIL.Image for example
```
```suggestion
})

# You can access the image with PIL.Image for example
```

osanseviero reviewed (commented):
Cool stuff! :fire:

Wauplin commented:
With https://github.com/huggingface/hub-docs/pull/1386 being merged, we have a clean first part to merge into the `new_api_docs` now. Let's not forget the few TODOs there.

osanseviero commented on docs/api-inference/tasks/image_to_image.md:
```diff
@@ -0,0 +1,63 @@
+
+</Tip>
+
+### Recommended models
```
As discussed before, we'll need to filter out models that are not warm. The one I would have here is https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0

Wauplin replied:
I thought twice about it and I don't know how we can do that. The `inference=warm` status can change over time but since the docs are static once generated, it might not be accurate when a user visits the page.

Wauplin replied:
I pushed https://github.com/huggingface/hub-docs/pull/1384/commits/d60825f8fc0fe3fa8ad5877772dac763ce972089 to fetch the inference status of each model. For now I haven't changed the templates for the "Recommended models" part. For `text-to-image` task it would be ok but for `image-to-image` task there are no warm models in the suggested ones.

Wauplin replied:
maybe https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0 should be listed on https://huggingface.co/tasks/image-to-image but that seems like a short term solution

Wauplin replied:
https://github.com/huggingface/hub-docs/pull/1396/files

osanseviero commented on docs/api-inference/tasks/text_to_image.md:
```diff
@@ -0,0 +1,116 @@
+| :--- | :--- | :--- |
+| **inputs** | _string, required_ | The input text data (sometimes called "prompt" |
+| **parameters** | _object, optional_ | Additional inference parameters for Text To Image |
+| **&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;guidance_scale** | _number, optional_ | For diffusion models. A higher guidance scale value encourages the model to generate images closely linked to the text prompt at the expense of lower image quality. |
```
Is the "diffusion models" clarification important? Only `diffusers` supports this task

Wauplin replied:
:man_shrugging: 

Not really no. I don't know where the original description comes from but we can update the specs in `huggingface.js` yes

osanseviero commented on scripts/api-inference/templates/specs_headers.handlebars:
```diff
@@ -0,0 +1,5 @@
+| :--- | :--- | :--- |
+| **authorization** | _string, optional_ | Authentication header in the form `'Bearer: hf_****'` when `hf_****` is a personal user access token with Inference API permission. You can generate one from [your settings page](https://huggingface.co/settings/tokens). |
+| **x-use-cache** | _boolean, optional, default to `true`_ | There is a cache layer on the inference API to speed up requests we have already seen. Most models can use those results as they are deterministic (meaning the outputs will be the same anyway). However, if you use a nondeterministic model, you can set this parameter to prevent the caching mechanism from being used, resulting in a real n ...
+| **x-wait-for-model** | _boolean, optional, default to `false`_ | If the model is not ready, wait for it instead of receiving 503. It limits the number of requests required to get your inference done. It is advised to only set this flag to true after receiving a 503 error, as it will limit hanging in your application to known places. Read more about model availability [here](../parameters#additi ...
```
```suggestion
| **x-wait-for-model** | _boolean, optional, default to `false`_ | If the model is not ready, wait for it instead of receiving 503. It limits the number of requests required to get your inference done. It is advised to only set this flag to true after receiving a 503 error, as it will limit hanging in your application to known places. Read more about model availability [here](../overview#eligibility]). |
```

osanseviero commented on scripts/api-inference/templates/specs_headers.handlebars:
```diff
@@ -0,0 +1,5 @@
+| Headers |   |    |
+| :--- | :--- | :--- |
+| **authorization** | _string, optional_ | Authentication header in the form `'Bearer: hf_****'` when `hf_****` is a personal user access token with Inference API permission. You can generate one from [your settings page](https://huggingface.co/settings/tokens). |
+| **x-use-cache** | _boolean, optional, default to `true`_ | There is a cache layer on the inference API to speed up requests we have already seen. Most models can use those results as they are deterministic (meaning the outputs will be the same anyway). However, if you use a nondeterministic model, you can set this parameter to prevent the caching mechanism from being used, resulting in a real n ...
```
```suggestion
| **x-use-cache** | _boolean, optional, default to `true`_ | There is a cache layer on the inference API to speed up requests we have already seen. Most models can use those results as they are deterministic (meaning the outputs will be the same anyway). However, if you use a nondeterministic model, you can set this parameter to prevent the caching mechanism from being used, resulting in a real new query. Read more about caching [here](../parameters#caching]). |
```

osanseviero commented on scripts/api-inference/scripts/generate.ts:
```diff
@@ -0,0 +1,301 @@
+import { snippets, PipelineType } from "@huggingface/tasks";
```
I'm skipping this file for now

osanseviero reviewed (commented):
Nice :fire:
